### PR TITLE
feat: support wan22 pipeline.

### PR DESCRIPTION
**Pull request by ethan686**
1. add the wan22 pipeline;
2. refactor the existed problem of other wan22 files; 
_  add eos token for sentencepiecetokenizer;_  ---- revert this change. 
4. curent percision compared with mindiesd with 40 steps : guidance_scale = 1.5, percision 0.9995 ; guidance_scale = 5.0, percision = 0.9689
5. current time consuming with 40 steps and 480*832 = 369.55s
6. supprt cfg parallel

### CODE REVIEW

File: xllm/models/dit/pipelines/pipeline_wan_i2v.h [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+            .view({1, num_channels_latents, 1, 1, 1})
+            .to(prepared_latents.device());
+
+    torch::Tensor latents_std = 1.0 / latents_std_raw;
+    prepared_latents = prepared_latents / latents_std;
+    prepared_latents = prepared_latents + latents_mean;
+    video = vae_->decode(prepared_latents.to(torch::kFloat32)).sample;
+    torch::save(
+        video,
+        "/export/home/weinan5/zjs/tensors_save_dir/cpp/vae_output_cpp.pt");
```
[the remark this answers is not in the document]
Comment (ethan686) [reply xllm/models/dit/pipelines/pipeline_wan_i2v.h]: delete the torch save.

File: xllm/models/dit/pipelines/pipeline_wan_i2v.h [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+class Wan2_2I2VPipelineImpl : public torch::nn::Module {
+ public:
+  Wan2_2I2VPipelineImpl(const DiTModelContext& context)
+      : parallel_args_(context.get_parallel_args()) {
+    options_ = context.get_tensor_options();
+    const auto& vae_args = context.get_model_args("vae");
+    zdim_ = vae_args.z_dim();
+    latents_mean_ = vae_args.vae_latents_mean();
+    latents_std_ = vae_args.vae_latents_std();
+    LOG(INFO) << "Pipeline loaded latents_mean size=" << latents_mean_.size()
```
[the remark this answers is not in the document]
Comment (ethan686) [reply xllm/models/dit/pipelines/pipeline_wan_i2v.h]: delete the log(info)

File: xllm/models/dit/pipelines/pipeline_wan_i2v.h [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+    register_module("transformer_2", transformer_2_);
+    register_module("umt5", umt5_);
+    register_module("scheduler", scheduler_);
+    register_module("video_processor_", video_processor_);
+  }
+
+  DiTForwardOutput forward(const DiTForwardInput& input) {
+    const auto& generation_params = input.generation_params;
+
+    auto seed = generation_params.seed > 0 ? generation_params.seed : 42;
```
[the remark this answers is not in the document]
Comment (ethan686) [reply xllm/models/dit/pipelines/pipeline_wan_i2v.h]: use int64_t instead.

File: xllm/models/dit/pipelines/pipeline_wan_i2v.h [hunk 4]
```diff
@@ -0,0 +1,684 @@
...
+    for (int64_t i = 0; i < batch_size; ++i) {
+      int64_t seq_len = seq_lens[i].item<int64_t>();
+      auto trimmed = prompt_embeds[i].slice(0, 0, seq_len);
+      int64_t padding_len = max_sequence_length - trimmed.size(0);
+      if (padding_len > 0) {
+        auto zeros =
+            torch::zeros({padding_len, trimmed.size(1)}, trimmed.options());
+        trimmed = torch::cat({trimmed, zeros}, 0);
+      }
+      trimmed_embeds.push_back(trimmed);
```
[the remark this answers is not in the document]
Comment (ethan686) [reply xllm/models/dit/pipelines/pipeline_wan_i2v.h]: fix.

File: xllm/models/dit/pipelines/pipeline_wan_i2v.h [hunk 5]
```diff
@@ -0,0 +1,684 @@
...
+    int64_t batch_size;
+    if (prompt_embeds.has_value()) {
+      batch_size = prompt_embeds_tensor.size(0);
+    } else if (prompt.has_value()) {
+      batch_size = prompt.value().size();
+    }
+
+    if (do_classifier_free_guidance) {
+      if (negative_prompt.has_value()) {
+        int prompt_size = negative_prompt.value().size();
```
[the remark this answers is not in the document]
Comment (ethan686) [reply xllm/models/dit/pipelines/pipeline_wan_i2v.h]: delete the unused variable.

File: xllm/models/dit/pipelines/pipeline_wan_i2v.h [hunk 6]
```diff
@@ -0,0 +1,684 @@
...
+
+        if (!timestep_input.defined()) {
+          timestep_input = t.expand(prepared_latents.size(0));
+        }
+      }
+      torch::Tensor noise_pred;
+      torch::Tensor noise_uncond;
+      if (do_classifier_free_guidance) {
+        if (FLAGS_cfg_size == 2) {
+          auto rank = parallel_args_.dit_cfg_group_->rank();
```
[the remark this answers is not in the document]
Comment (ethan686) [reply xllm/models/dit/pipelines/pipeline_wan_i2v.h]: use int32_t instead.

File: xllm/models/dit/pipelines/pipeline_wan_i2v.h [hunk 7]
```diff
@@ -0,0 +1,684 @@
...
+  torch::Tensor get_t5_prompt_embeds(std::vector<std::string>& prompt,
+                                     int64_t num_videos_per_prompt = 1,
+                                     int64_t max_sequence_length = 512,
+                                     const std::string& save_tag = "") {
+    int64_t batch_size = prompt.size();
+
+    std::vector<std::vector<int32_t>> text_input_ids;
+    text_input_ids.reserve(batch_size);
+
+    for (int i = 0; i < prompt.size(); i++) {
```
[the remark this answers is not in the document]
Comment (ethan686) [reply xllm/models/dit/pipelines/pipeline_wan_i2v.h]: fix.

File: xllm/models/dit/pipelines/pipeline_wan_i2v.h [hunk 8]
```diff
@@ -0,0 +1,684 @@
...
+                                 t)
+                                    .flatten();
+        timestep_input =
+            temp_ts.unsqueeze(0).expand({prepared_latents.size(0), -1});
+      } else {
+        latent_model_input =
+            torch::cat({prepared_latents, latent_condition}, 1);
+        latent_model_input = latent_model_input.to(prepared_latents.dtype());
+
+        if (!timestep_input.defined()) {
```
[the remark this answers is not in the document]
Comment (ethan686) [reply xllm/models/dit/pipelines/pipeline_wan_i2v.h:507]: if not changed, also fine. since you mention this and with {} is eaiser to understand, so fix.

File: xllm/core/framework/dit_model_loader.cpp [hunk 9]
```diff
@@ -441,6 +441,18 @@ bool DiTFolderLoader::load_tokenizer_args(
...
       tokenizer_args_.vocab_file() = tokenizer_json_path;
+    } else {
+      // T5-family models use SentencePiece vocab. Probe for the vocab file
+      // (spiece.model for UMT5, tokenizer.model for T5) in the model directory.
+      for (const auto& candidate : {"spiece.model", "tokenizer.model"}) {
+        const std::string candidate_path =
+            model_weights_path_ + "/" + candidate;
+        if (std::filesystem::exists(candidate_path)) {
+          tokenizer_args_.tokenizer_type() = "sentencepiece";
+          tokenizer_args_.vocab_file() = candidate;
```
[the remark this answers is not in the document]
Comment (ethan686) [reply xllm/core/framework/dit_model_loader.cpp]: fix

**Review by XuZhang99 — DISMISSED** [review]

File: xllm/core/framework/model/model_args.h
Comment (xiao-yu-chen) [inline xllm/core/framework/model/model_args.h]: similar fields are redefined. refer to line 521.

File: xllm/core/framework/request/dit_request_params.cpp [hunk 10]
```diff
@@ -215,7 +149,48 @@ DiTRequestParams::DiTRequestParams(const proto::ImageGenerationRequest& request,
...
+    }
+    input_params.images.reserve(input.images().size());
+    for (const auto& image : input.images()) {
+      torch::Tensor tensor;
+      if (!decode_base64_image(image, tensor)) {
+        continue;
+      }
+      input_params.images.emplace_back(std::move(tensor));
+    }
+    if (input.has_condition_image()) {
```
Comment (xiao-yu-chen) [inline xllm/core/framework/request/dit_request_params.cpp]: condition_image still exist ？

File: xllm/models/dit/encoders/umt5_encoder.h [hunk 11]
```diff
@@ -235,6 +283,7 @@ class UMT5EncoderModelImpl final : public torch::nn::Module {
   }
 
  private:
+  bool is_wan_style_ = false;
```
Comment (xiao-yu-chen) [inline xllm/models/dit/encoders/umt5_encoder.h]: wan cannot reused umt5 directly？

File: xllm/core/framework/dit_model_loader.cpp [hunk 12]
```diff
@@ -441,6 +441,18 @@ bool DiTFolderLoader::load_tokenizer_args(
     if (!is_t5) {
       tokenizer_args_.tokenizer_type() = "fast";
       tokenizer_args_.vocab_file() = tokenizer_json_path;
+    } else {
+      // T5-family models use SentencePiece vocab. Probe for the vocab file
```
Comment (xiao-yu-chen) [inline xllm/core/framework/dit_model_loader.cpp]: The logic for calculating the tokenizer type is a bit confusing

File: xllm/models/dit/pipelines/pipeline_wan_i2v.h [hunk 13]
```diff
@@ -0,0 +1,678 @@
...
+#include "models/dit/autoencoders/autoencoder_kl_wan.h"
+#include "models/dit/encoders/umt5_encoder.h"
+#include "models/dit/processors/vae_video_processor.h"
+#include "models/dit/schedulers/uni_pc_multi_step_scheduler.h"
+#include "models/dit/transformers/transformer_wan.h"
+#include "models/model_registry.h"
+
+namespace xllm {
+
+class Wan2_2I2VPipelineImpl : public torch::nn::Module {
```
Comment (xiao-yu-chen) [inline xllm/models/dit/pipelines/pipeline_wan_i2v.h]: 2_2 is not needed

File: xllm/models/dit/transformers/transformer_wan.h
Comment (xiao-yu-chen) [inline xllm/models/dit/transformers/transformer_wan.h]: The class name needs to be aligned with diffusers

File: xllm/models/dit/pipelines/pipeline_wan_i2v.h (hunk 13 above)
Comment (xiao-yu-chen) [reply xllm/models/dit/pipelines/pipeline_wan_i2v.h]: The class name needs to be aligned with diffusers

File: xllm/core/framework/model/model_args.h
Comment (ethan686) [reply xllm/core/framework/model/model_args.h]: delete 8 duplicate vae args， reuse the args in qwen_image_edit

File: xllm/core/framework/request/dit_request_params.cpp (hunk 10 above)
Comment (ethan686) [reply xllm/core/framework/request/dit_request_params.cpp]: delete all the condition image logic.

File: xllm/models/dit/encoders/umt5_encoder.h (hunk 11 above)
Comment (ethan686) [reply xllm/models/dit/encoders/umt5_encoder.h]: longcat's umt5 and wan22's t5 has one differnt weight name, and different logic in attention mask

Comment (ethan686) [reply xllm/models/dit/encoders/umt5_encoder.h]: still, the flag can be removed.  i remove the flag with a small umt5 logic change  to support both scenarios.

File: xllm/models/dit/pipelines/pipeline_wan_i2v.h (hunk 13 above)
Comment (ethan686) [reply xllm/models/dit/pipelines/pipeline_wan_i2v.h]: fix, aligened with diffusers.

File: xllm/models/dit/transformers/transformer_wan.h
Comment (ethan686) [reply xllm/models/dit/transformers/transformer_wan.h]: this module is useless, delete. only keep the WanTransformer3DModel which is same with diffuers.

File: xllm/core/framework/dit_model_loader.cpp (hunk 12 above)
Comment (ethan686) [reply xllm/core/framework/dit_model_loader.cpp]: refactored

File: xllm/core/framework/dit_model_loader.cpp
Comment (xiao-yu-chen) [inline xllm/core/framework/dit_model_loader.cpp]: delete ?

File: xllm/models/dit/encoders/umt5_encoder.h [hunk 14]
```diff
@@ -211,6 +224,12 @@ class UMT5EncoderModelImpl final : public torch::nn::Module {
                           "encoder.embed_tokens.weight",
                           embed_tokens_->weight,
                           is_embed_tokens_loaded_);
+      if (!is_embed_tokens_loaded_) {
```
Comment (xiao-yu-chen) [inline xllm/models/dit/encoders/umt5_encoder.h:224]: if condition  is needed ?

File: xllm/models/dit/encoders/umt5_encoder.h [hunk 15]
```diff
@@ -183,11 +193,16 @@ class UMT5EncoderModelImpl final : public torch::nn::Module {
...
-  torch::Tensor forward(const torch::Tensor& input_ids) {
+  torch::Tensor forward(
+      const torch::Tensor& input_ids,
+      const std::optional<torch::Tensor>& attention_mask = std::nullopt) {
     torch::Tensor hidden_states = embed_tokens_->forward(input_ids);
-    // UMT5: no position_bias passed between blocks — each block recomputes it.
+
     for (size_t i = 0; i < layers_.size(); ++i) {
-      auto layer_outputs = layers_[i]->forward(hidden_states);
+      auto layer_outputs =
```
Comment (xiao-yu-chen) [inline xllm/models/dit/encoders/umt5_encoder.h]: ：？ is needed？

Comment (ethan686) [reply xllm/models/dit/encoders/umt5_encoder.h]: yes，no need， delete

File: xllm/core/framework/dit_model_loader.cpp
Comment (ethan686) [reply xllm/core/framework/dit_model_loader.cpp]: deleted

File: xllm/models/dit/encoders/umt5_encoder.h (hunk 14 above)
Comment (ethan686) [reply xllm/models/dit/encoders/umt5_encoder.h:224]: yes, its needed. longcat and wan22's umt5 has different weights here.

**Review by xiao-yu-chen — APPROVED** [review]

**Review by Dragonliu2018 — APPROVED** [review]
LGTM

**Review by XuZhang99 — APPROVED** [review]
